tests/extension/oneapi_non_uniform_groups/opportunistic_group_api.cpp

```diff
@@ -0,0 +1,182 @@
+/*******************************************************************************
+//
+//  SYCL 2020 Conformance Test Suite
+//
+//  Copyright (c) 2024 The Khronos Group Inc.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+*******************************************************************************/
+
+#include "../../common/common.h"
+
+namespace non_uniform_groups::tests {
+
+struct checks {
+  enum {
+    get_group_id,
+    get_local_id,
+    get_group_range,
+    get_local_range,
+    get_group_linear_id,
+    get_local_linear_id,
+    get_group_linear_range,
+    get_local_linear_range,
+    leader,
+    COUNT,
+  };
+};
+
+TEST_CASE("Test for opportunistic_group apis.", "[oneapi_non_uniform_groups]") {
+#ifndef SYCL_EXT_ONEAPI_NON_UNIFORM_GROUPS
+  SKIP("SYCL_EXT_ONEAPI_NON_UNIFORM_GROUPS is not defined");
+#else
+  namespace oneapi_ext = sycl::ext::oneapi::experimental;
+  using opportunistic_group_t = oneapi_ext::opportunistic_group;
+  using CheckResults = bool[checks::COUNT];
+
+  constexpr size_t num_items = 64;
+
+  sycl::buffer<CheckResults, 1> results_buffer{num_items};
+
+  auto q = sycl_cts::util::get_cts_object::queue();
+
+  if (!q.get_device().has(sycl::aspect::ext_oneapi_opportunistic_group)) {
+    SKIP("Device does not support opportunistic_group.");
+  }
+
+  q.submit([&](sycl::handler& cgh) {
+    sycl::accessor acc{results_buffer, cgh, sycl::write_only};
+
+    sycl::nd_range<1> nd_range{sycl::range<1>{num_items},
+                               sycl::range<1>{num_items}};
+
+    cgh.parallel_for<struct OpportunisticGroupAPI>(
+        nd_range, [=](sycl::nd_item<1> it) {
+          auto& results = acc[it.get_global_id()];
+
+          size_t sub_group_size = it.get_sub_group().get_local_range().size();
+
+          auto opportunistic =
+              oneapi_ext::this_kernel::get_opportunistic_group();
```

Comment:
Should this be checked?
```
Each call to `get_opportunistic_group()` returns a different group.
```

Reply:
That is a good question. @Pennycook - What is the intention here? Do the groups have [common by-value semantics](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:byval-semantics) and if so, does this make the guarantee that `get_opportunistic_group() != get_opportunistic_group()`, which in turn means the groups will never have the same members values?

Reply:
That is a very good question.

The intention of the wording was really to say that the members _may_ be different, and that in cases where the members _are_ different, the group objects may be different.  Without thinking of any specific implementation: if the group members are different, the representation (e.g., as a mask) will probably be different; even if the group members are the same, I can imagine situations where the different group objects would still have dedicated scratchpad memory or barriers.

I think the simplest thing to do for now is to guarantee that equality returns false.  It's probably more useful to have a comparison that is always useless than to try and define the cases in which equality should/could return true.  I'd prefer not to provide a comparison operator at all, but I've heard suggestions recently that this would break a lot of things.

Reply:
> I think the simplest thing to do for now is to guarantee that equality returns false. It's probably more useful to have a comparison that is always useless than to try and define the cases in which equality should/could return true. I'd prefer not to provide a comparison operator at all, but I've heard suggestions recently that this would break a lot of things.

Are you saying that `opportunistic_group` needs to have a comparison operator?  If you are saying that things would break if there is no comparison operator, it seems surprising that those things wouldn't also break if the comparison operator always returned false.

Are you saying that the comparison operator would return false even when comparing an object with a copy?  For example:

```
opportunistic_group og1 = /*...*/;
opportunistic_group og2 = og1;
bool same = (og1 == og2);  // returns false?
```

Reply:
> Are you saying that `opportunistic_group` needs to have a comparison operator? If you are saying that things would break if there is no comparison operator, it seems surprising that those things wouldn't also break if the comparison operator always returned false.

I don't have complete information here.  This came up in a discussion we were having about some of the other types (e.g., `nd_item`), and my takeaway was that that if we were to remove certain operators then the classes wouldn't play nicely with certain type traits.  I might have gotten the wrong end of the stick.

> Are you saying that the comparison operator would return false even when comparing an object with a copy? For example:
> 
> ```
> opportunistic_group og1 = /*...*/;
> opportunistic_group og2 = og1;
> bool same = (og1 == og2);  // returns false?
> ```

Yes, that's what I was suggesting.

I'm also open to defining a real comparison operator for groups, but I don't know what the criteria should be.  I can see an argument that it should be based on which work-items are in the group (which would allow comparisons across groups of different types) but can also think of cases where that could lead to surprising results (e.g., just because two groups have the same work-items in them doesn't necessarily mean they can be used interchangably).

Reply:
If this is an open issue for groups in general, could we leave it out of these tests for now and address it for all groups when it has been discussed and consensus has been reached on how it should work?